sygnal/webpushpushkin.py

```diff
@@ -0,0 +1,167 @@
+# -*- coding: utf-8 -*-
+# Copyright 2021 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import logging
+import time
+from io import BytesIO
+from json import JSONDecodeError
+
+from pywebpush import webpush, WebPushException
+from py_vapid import Vapid
+from opentracing import logs, tags
+from prometheus_client import Counter, Gauge, Histogram
+from twisted.enterprise.adbapi import ConnectionPool
+from twisted.internet.defer import DeferredSemaphore
+from twisted.web.client import FileBodyProducer, HTTPConnectionPool, readBody
+from twisted.web.http_headers import Headers
+
+from sygnal.exceptions import (
+    NotificationDispatchException,
+    TemporaryNotificationDispatchException,
+)
+from sygnal.helper.context_factory import ClientTLSOptionsFactory
+from sygnal.helper.proxy.proxyagent_twisted import ProxyAgent
+from sygnal.utils import NotificationLoggerAdapter, twisted_sleep
+
+from .exceptions import PushkinSetupException
+from .notifications import ConcurrencyLimitedPushkin
+
+logger = logging.getLogger(__name__)
+
+MAX_TRIES = 3
+RETRY_DELAY_BASE = 10
+MAX_BYTES_PER_FIELD = 1024
+
+DEFAULT_MAX_CONNECTIONS = 20
+
+
+class WebpushPushkin(ConcurrencyLimitedPushkin):
+    """
+    Pushkin that relays notifications to Google/Firebase Cloud Messaging.
+    """
+
+    UNDERSTOOD_CONFIG_FIELDS = {
+        "type",
+        "max_connections",
+        "vapid_private_key",
+        "vapid_contact_email",
+    } | ConcurrencyLimitedPushkin.UNDERSTOOD_CONFIG_FIELDS
+
+    def __init__(self, name, sygnal, config):
+        super(WebpushPushkin, self).__init__(name, sygnal, config)
+
+        nonunderstood = set(self.cfg.keys()).difference(self.UNDERSTOOD_CONFIG_FIELDS)
+        if len(nonunderstood) > 0:
```

Comment:
Might be more clearly expressed as:

```python
nonunderstood = self.cfg.keys() - self.UNDERSTOOD_CONFIG_FIELDS
if nonunderstood:
    ...
```

But these follow patterns elsewhere in Sygnal so 🤷